Stats: add passing select UTM parameters

trakos opened the pull request:
## Proposed changes:
This PR adds select UTM parameters to the pixel requests send to wp.com, so that we can show more statistics to site owners.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
panfyZ-37z-p2

## Does this pull request change what data or activity we track or use?
It adds additional parameters to tracking, but we already mention tracking "visiting URL", which parameters are part of.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Make sure you are logged out
* Add `?utm_source=a_source&utm_medium=a_medium` to any page on your test website
* Verify that the request to pixel.wp.com includes the UTM parameters
* Add a plugin for AMP support, visit any post with `?amp=1&utm_source=test` added to URL, verify that the pixel URL includes UTM parameter

trakos committed:
Stats: add passing select UTM parameters

jeherve commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -61,7 +73,16 @@ public static function build_view_data() {
-		return compact( 'v', 'blog', 'post', 'tz', 'srv' );
+		$view_data = compact( 'v', 'blog', 'post', 'tz', 'srv' );
+		foreach ( self::TRACKED_UTM_PARAMETERS as $utm_parameter ) {
+			// phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
When we ignore specific PHPCS rules, I often recommend adding a comment explaining why. It makes things a bit easier later on, when coming back to a file that was edited a long time ago ; it feels safer knowing that we added the exception knowingly.

PHPCS allows adding comments like so:

```suggestion
			// phpcs:ignore WordPress.Security.NonceVerification.Recommended -- we clear the parameter as we output it below.
```

trakos replied:
Thank you for the suggestion! I've added a comment explaining why. UTM parameters are coming from external websites by design, so we don't control them and can't add a nonce here.

jeherve commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -61,7 +73,16 @@ public static function build_view_data() {
+		foreach ( self::TRACKED_UTM_PARAMETERS as $utm_parameter ) {
+			// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+			if ( isset( $_GET[ $utm_parameter ] ) && is_scalar( $_GET[ $utm_parameter ] ) ) {
+				// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.NonceVerification.Recommended
+				$view_data[ $utm_parameter ] = substr( (string) $_GET[ $utm_parameter ], 0, 255 );
```
What would you think about adding some of the recommended sanitization methods here? UTM parameter values are limited on Google's end as well I believe, so why not do the same on our end? It would also help avoid issues like PHP notices being logged when someone tries to do something like `?utm_source=<script>alert( 'hello' )</script>`.

```
PHP Notice:  Function wp_add_inline_script was called incorrectly. Do not pass &lt;script&gt; tags to wp_add_inline_script(). Please see <a>Debugging in WordPress</a> for more information. (This message was added in version 4.5.0.) in /wp-includes/functions.php on line 6031
```

```suggestion
				// phpcs:ignore WordPress.Security.NonceVerification.Recommended -- We're sanitizing the value.
				$view_data[ $utm_parameter ] = substr( sanitize_textarea_field( wp_unslash( $_GET[ $utm_parameter ] ) ), 0, 255 );
```

trakos replied:
I've added the sanitization. It does limit the valid UTM parameters we can handle slightly, but it shouldn't come up in a real-world scenarios too often.

By the way, would you know why `Tracking_Pixel::stats_array_to_string` didn't use `json_encode` directly? I've changed it now that you've highlighted the importance of escaping the data properly. I don't think `addslashes` was sufficient there, and it wasn't checking keys, which could have been overlooked when adding more parameters.

gibrown replied:
> would you know why Tracking_Pixel::stats_array_to_string didn't use json_encode directly?

Just a fun guess because I was curious... json_encode() was added in php5.2 in late 2006. The original stats_array() function on wpcom in mu-plugins was written in Feb 2007!

trakos replied:
Oh, thanks for checking! I didn't realize it was moved from an old code, that would definitely explain it. I think it's worth changing it to json_encode then, just in case we change array keys in the future.

kangzj reviewed (commented):
Thanks for working on this. It works well for me 👍 

@jeherve's https://github.com/Automattic/jetpack/pull/34431/files#r1424316039 seems important to address tho.

trakos committed:
Stats: sanitize utm parameters

trakos committed:
Improve encoding data for JS

trakos committed:
Fix tests

trakos commented:
Thank you for reviews @kangzj and @jeherve! I've applied the suggestions and changed the serialization in `build_stats_details` method.

Would you know why `Tracking_Pixel::stats_array_to_string` didn't use `json_encode` before? I've changed it now that you've highlighted the importance of escaping the data properly. I don't think `addslashes` was sufficient there, and it wasn't checking keys, which could have been overlooked when adding more parameters.

trakos committed:
Update composer.lock

trakos committed:
Fix for batcache

trakos commented:
When working on wpcom counter-part I've noticed that we need to account for certain cache configs on Atomic: panfyZ-37z-p2#comment-7723, so I've updated this PR.

sotirispl reviewed (approved):
I tested it and it works as described. Thank you for working on this!
